src/backends/kimchi/mod.rs

```diff
@@ -384,7 +385,24 @@ impl Backend for KimchiVesta {
                 witness_row[col] = val;
             }
 
-            // check if the row makes sense
+            witness.push(witness_row);
+        }
+
+        // compute public output at last
+        let mut public_outputs = vec![];
+
+        for (var, rows_cols) in public_outputs_vars {
+            let val = self.compute_var(witness_env, var)?;
+            for (row, col) in rows_cols {
+                witness[row][col] = val;
+            }
+            public_outputs.push(val);
+        }
+
+        // sanity check the witness
+        for (row, (gate, witness_row, debug_info)) in
+            izip!(self.gates.iter(), &witness, &self.debug_info).enumerate()
+        {
```

Comment:
btw what is this change related to in this file?

Reply:
The `generate_witness` function was assuming there is only a `PublicOutput` var in the circuit. But after constraining the output to the result of circuit, this public output cell will appear twice at least in the circuit:
 - One time as the public output constraint (as part of the public entry), which will be cancelled by the verifier
 - Second time as the part of the constraint with the result of the circuit

So here we need to have a hash map to record where are the places this output cell appears in circuit. With this location info, the output value calculated at the last via `compute_var` can be put to the correct `<row, col>`.

Reply:
oh right... I was thinking about it like we're using the permutation for `assert_eq`, but we still don't :D (cf https://github.com/zksecurity/noname/issues/55)